### PR TITLE
Extend HolographicSeries from Timeseries

### LINKED ISSUE #4: Cannot extend OptogeneticSeries to have 2D data
**Extension specification:**
https://github.com/catalystneuro/ndx-holographic-stimulation/blob/82fba2b7763a06ad35563252e05171fed269c372/src/spec/create_extension_spec.py#L63-L77

When running `test_example_usage()`, it displays this error message:
```python
ValueError: CustomClassGenerator.set_init.<locals>.__init__: incorrect shape for 'data' (got '(100, 12)', expected '(None,)')
```
It still expects the data shape of the OptogeneticSeries (None,), even if (None, None) was defined as the new data shape.
Extending HolographicSeries from TimeSeries resolves the error.

### PR DESCRIPTION
**Pull request by alessandratrapani**
Fix #4 
- [x] extend HolographicSeries from TimeSeries
- [x] extend HolographicStimulusPattern from LabMetadata
- [x] implement SpiralScanning table
- [x] implement TemporalFocusing table
- [x] unit test
- [ ] example files that use this structure from both the mouse V1 and Pinto lab

### CODE REVIEW

File: spec/ndx-holographic-stimulation.extensions.yaml [hunk 1]
```diff
@@ -22,24 +22,30 @@ groups:
...
   datasets:
   - name: data
-    neurodata_type_inc: VectorData
     dtype: float
     shape:
-    - null
-    - null
+    - - null
+    - - null
+      - null
```
Comment (CodyCBakerPhD) [inline spec/ndx-holographic-stimulation.extensions.yaml]: As discussed, this should simply be 2D at all times

Reference for RoiResponseSeries schema: https://github.com/NeurodataWithoutBorders/nwb-schema/blob/ec0a87979608c75a785d3a42f61e3366846ed3c2/core/nwb.ophys.yaml#L79-L81

File: spec/ndx-holographic-stimulation.extensions.yaml [hunk 2]
```diff
@@ -22,24 +79,32 @@ groups:
     doc: link to the holographic stimulus pattern
     quantity: '?'
 - neurodata_type_def: HolographicSeries
-  neurodata_type_inc: OptogeneticSeries
```
Comment (weiglszonja) [inline spec/ndx-holographic-stimulation.extensions.yaml:25]: Opened this at https://github.com/NeurodataWithoutBorders/nwb-schema/issues/556 to get some help why we can't extend `OptogeneticSeries`

File: spec/ndx-holographic-stimulation.extensions.yaml [hunk 3]
```diff
@@ -22,24 +79,32 @@ groups:
...
-    neurodata_type_inc: VectorData
-    dtype: float
+    dtype: numeric
+    dims:
+    - num_times
+    - num_rois
     shape:
     - null
     - null
     doc: The data values. May be 1D or 2D. The first dimension must be time. The optional
```
Comment (weiglszonja) [inline spec/ndx-holographic-stimulation.extensions.yaml]: This doc should be updated if we restrict to 2d only

File: spec/ndx-holographic-stimulation.extensions.yaml [hunk 4]
```diff
@@ -22,24 +79,32 @@ groups:
...
     quantity: '?'
 - neurodata_type_def: HolographicSeries
-  neurodata_type_inc: OptogeneticSeries
+  neurodata_type_inc: TimeSeries
   doc: An extension of OptogeneticSeries to include the holographic representation
     for the stimulus.
+  attributes:
+  - name: unit
+    dtype: text
+    default_value: W
```
Comment (weiglszonja) [inline spec/ndx-holographic-stimulation.extensions.yaml]: this should follow core-schema naming:
https://github.com/NeurodataWithoutBorders/nwb-schema/blob/ec0a87979608c75a785d3a42f61e3366846ed3c2/core/nwb.ogen.yaml#L16

File: src/pynwb/tests/test_holographic_stimulation.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            imaging_plane=self.imaging_plane, n_rois=self.n_rois, nwbfile=self.nwbfile
+        )
+
+        self.roi_table_region = self.plane_segmentation.create_roi_table_region(
+            region=[0, 1], description="the first of two ROIs"
+        )
+        # metadata for holographic series
+        self.series_name = "HolographicSeries"
+        self.series_description = "Holographic stimulus on 2 rois"
+        self.unit = "W"
```
Comment (weiglszonja) [inline src/pynwb/tests/test_holographic_stimulation.py]: https://github.com/NeurodataWithoutBorders/nwb-schema/blob/ec0a87979608c75a785d3a42f61e3366846ed3c2/core/nwb.ogen.yaml#L16

File: src/spec/create_extension_spec.py [hunk 6]
```diff
@@ -62,35 +150,50 @@ def main():
...
         doc=(
             "An extension of OptogeneticSeries to include the holographic representation for "
             "the stimulus."
         ),
+        attributes=[
+            NWBAttributeSpec(
+                name="unit",
+                doc="SI unit of data",
+                dtype="text",
+                default_value="W",
```
Comment (weiglszonja) [inline src/spec/create_extension_spec.py]: https://github.com/NeurodataWithoutBorders/nwb-schema/blob/ec0a87979608c75a785d3a42f61e3366846ed3c2/core/nwb.ogen.yaml#L16

File: .gitignore [hunk 7]
```diff
@@ -103,6 +103,19 @@ venv.bak/
...
+!.vscode/tasks.json
+!.vscode/launch.json
+!.vscode/extensions.json
+!.vscode/*.code-snippets
+
+# Local History for Visual Studio Code
+.history/
+
+# Built Visual Studio Code Extensions
+*.vsix
```
Comment (weiglszonja) [inline .gitignore]: Is this relevant to this PR? I think no need to push these changes to .gitignore

Comment (CodyCBakerPhD) [reply .gitignore]: This is essentially the first full PR for this repo so either we get this in or splinter it as a separate PR to be expedited

Comment (weiglszonja) [reply .gitignore]: Ok if you feel this is needed it can go in this one.

File: spec/ndx-holographic-stimulation.extensions.yaml (hunk 2 above)
Comment (alessandratrapani) [reply spec/ndx-holographic-stimulation.extensions.yaml:25]: Thank you so much!

File: src/pynwb/tests/test_holographic_stimulation.py [hunk 8]
```diff
@@ -48,7 +49,7 @@ def setUp(self) -> None:
         # metadata for holographic series
         self.series_name = "HolographicSeries"
         self.series_description = "Holographic stimulus on 2 rois"
-        self.unit = "W"
+        self.unit = "watts"
```
Comment (weiglszonja) [inline src/pynwb/tests/test_holographic_stimulation.py:54]: thank you!

File: .gitignore (hunk 7 above)
Comment (pauladkisson) [inline .gitignore]: Please add these to the global .gitignore file: [catalystneuro/.github/.gitignore](https://github.com/catalystneuro/.github)

**Review by pauladkisson — CHANGES_REQUESTED** [review]
Tests and extension overall look very good! k

After thinking about it some more, if we put the rois in the holographicseries then we should be able to make multiple references to a subset of the rois in the plane segmentation table with only a single plane segmentation object, right? Earlier I was worried that we would need to create many plane segmentation objects, but since rois is a dynamic table region, that shouldn't be the case. If we keep rois in the holographic stimulus site, that we would run into the issue that the data array is much bigger than it needs to be if you only stimulates a subset of rois. Sorry for flip-flopping on this issue, I just wanted to consider it and make sure we get it right.
